### DISCUSSION TITLE
OpenApi: How to change model name?

### DISCUSSION THREAD

**Discussion by Thijmen** [dotnet/aspnetcore#58999] [Q&A] [answered]
Hi there,

First of all; I am loving the Dotnet 9 updates! I am diving right in and I have some difficulties regarding renaming model names.

Say the given scenario:

```c#
public static class SomeCommand {
    public record Command(string Name);
}

public static class SomeOtherCommand {
    public record Command(string Name);
}
````

I use these records as bodies in my APIs. When using the latest dotnet 9 openapi services, the model names end up as "Command", "Command2", "Command3" etc. Ideally, I would like some filter to change these.

This is what I have as of now:
```c#
public class CommandSchemaNameTransformer : IOpenApiSchemaTransformer
{
    public Task TransformAsync(OpenApiSchema schema, OpenApiSchemaTransformerContext context, CancellationToken cancellationToken)
    {
        if (context.JsonPropertyInfo == null)
        {
            var type = context.JsonTypeInfo.Type;
            var schemaType = typeof(SomeOtherCommand.Command);
            if (type == schemaType)
            {
                var fullName = schemaType.FullName;
                if (!string.IsNullOrWhiteSpace(fullName) && fullName.Contains("+"))
                {
                    var parts = fullName.Split('.');
                    var lastPart = parts[^1];
                    var commandParts = lastPart.Split('+');
                    schema.Title = string.Join("", commandParts);
                }
            }
        }
        
        return Task.CompletedTask;
    }
} 
```

This does update the title, but not the schema name:

<img width="357" alt="image" src="https://github.com/user-attachments/assets/42b1c39e-edfa-4941-8101-e0761081061e">

Is what I want actually achievable?

**Comment by martincostello** [comment, marked as answer]
I believe that the `OpenApiOptions.CreateSchemaReferenceId` property is the extension point for this: https://github.com/dotnet/aspnetcore/blob/eb68e016a554b4da50d7fb0aeffe897cfabf36c7/src/OpenApi/src/Services/OpenApiOptions.cs#L52-L59

**Reply by Thijmen** [reply]
Hi @martincostello, thanks for getting back with me.

I've added it as an annotation in the same transformer as provided above;
```c#
schema.Annotations["x-schema-id"] = schema.Title;
```

Is this something that you would recommend or would you say I'd look in the CreateSchemaReferenceId?

**Reply by martincostello** [reply]
I would try the delegate property first, as it seems to be explicitly documented to be for that, rather than obliquely via a key in an annotation.

**Reply by Thijmen** [reply]
It is however used here too; 
https://github.com/dotnet/aspnetcore/blob/eb68e016a554b4da50d7fb0aeffe897cfabf36c7/src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs#L217

Let me take as swing at the delegate property.

**Reply by Thijmen** [reply]
@martincostello Too bad the `.GetSchemaReferenceId` method is not available for me. I'd like to only make changes if my class contains "Command", so I leave most stuff "per default". Any other suggestion here?

**Reply by martincostello** [reply]
You could check the name and if it isn't command you could use their default: https://github.com/dotnet/aspnetcore/blob/eb68e016a554b4da50d7fb0aeffe897cfabf36c7/src/OpenApi/src/Services/OpenApiOptions.cs#L26

**Reply by Thijmen** [reply]
You are totally right, oversight from my side. Been busy upgrading a project since last night. Maybe time for some coffee :D Thanks for the help @martincostello, truly appreciated. Cheers!

**Comment by petercinibulk** [comment]
@Thijmen Any chance you could share what your final solution was for changing a model name?
